chainerx_cc/chainerx/routines/normalization.cc

```diff
@@ -117,43 +273,70 @@ Array BatchNorm(
         Scalar eps,
         Scalar decay,
         const OptionalAxes& axis) {
+    // Preprocess inputs.
     PreprocessBatchNormResult result = PreprocessBatchNorm(x, gamma, beta, running_mean, running_var, axis);
-    std::shared_ptr<BatchNormForwardBackward> fb =
-            x.device().GetBatchNormForwardBackward(result.mean, result.var, eps, decay, result.sorted_axis);
-
     const Array& gamma_reshaped = result.gamma;
     const Array& beta_reshaped = result.beta;
+    const Array& mean_reshaped = result.mean;
+    const Array& var_reshaped = result.var;
+    const Axes& sorted_axis = result.sorted_axis;
+
+    Device& device = x.device();
+
+    // Compute forward.
+    Array out = EmptyLike(x, device);
+    std::shared_ptr<BatchNormState> state{};
+    {
+        NoBackpropModeScope scope{};
+        std::tuple<Array, std::shared_ptr<BatchNormState>> batch_norm_result = device.backend().CallOp<BatchNormOp>(
+                x.AsGradStopped(),
+                gamma_reshaped.AsGradStopped(),
+                beta_reshaped.AsGradStopped(),
+                mean_reshaped,
+                var_reshaped,
+                eps,
+                decay,
+                sorted_axis,
+                true,
+                out);
+        state = std::get<1>(batch_norm_result);
+    }
+    CHAINERX_ASSERT(state != nullptr);
 
-    Array out = fb->Forward(x.AsGradStopped(), gamma_reshaped.AsGradStopped(), beta_reshaped.AsGradStopped());
-    CHAINERX_ASSERT(out.dtype() == x.dtype());
     internal::MakeViewForForwardBackwardOutput(out);
 
     BackwardBuilder bb{"batch_norm", {x, gamma_reshaped, beta_reshaped}, {out}};
     if (BackwardBuilder::Target bt = bb.CreateTarget({0, 1, 2})) {
-        bt.Define([fb = std::move(fb), x_tok = bb.RetainInput(0), gamma_tok = bb.RetainInput(1), eps, sorted_axis = result.sorted_axis](
-                          BackwardContext& bctx) {
+        bt.Define([state = std::move(state),
+                   x_tok = bb.RetainInput(0),
+                   gamma_tok = bb.RetainInput(1),
+                   eps,
+                   sorted_axis,
+                   beta_shape = beta_reshaped.shape(),
+                   beta_dtype = beta_reshaped.dtype()](BackwardContext& bctx) mutable {
```

Comment:
Why `mutable`?

Reply:
Thanks, removed it. I think I used it before while mutating the captured state but it is not needed.